src/webgl/p5.RendererGL.js

```diff
@@ -1097,76 +1055,107 @@ p5.RendererGL = class RendererGL extends p5.Renderer {
-      this.filterShader.setUniform('tex0', tmp);
-      pg.clear();
-      pg.rect(-this.width / 2, -this.height / 2, this.width, this.height);
-
-      // read back to temp buffer
-      tmp.clear();
-      tmp.image(pg, -this.width / 2, -this.height / 2);
+      // Horiz pass: draw `target` to `tmp`
+      tmp.draw(() => {
+        this.filterShader.setUniform('direction', [1, 0]);
+        this.filterShader.setUniform('tex0', target);
+        this._pInst.clear();
+        this._pInst.shader(this.filterShader);
+        this._pInst.rect(-target.width / 2,
+          -target.height / 2, target.width, target.height);
+      });
 
-      // vert pass
-      this.filterShader.setUniform('direction', [0, 1]);
-      this.filterShader.setUniform('tex0', tmp);
-      pg.clear();
-      pg.rect(-this.width / 2, -this.height / 2, this.width, this.height);
+      // Vert pass: draw `tmp` to `fbo`
+      fbo.draw(() => {
+        this.filterShader.setUniform('direction', [0, 1]);
+        this.filterShader.setUniform('tex0', tmp);
+        this._pInst.clear();
+        this._pInst.shader(this.filterShader);
+        this._pInst.rect(-target.width / 2,
+          -target.height / 2, target.width, target.height);
+      });
 
       this._pInst.pop();
     }
     // every other non-blur shader uses single pass
     else {
-      pg.shader(this.filterShader);
-      this.filterShader.setUniform('tex0', this);
-      this.filterShader.setUniform('texelSize', texelSize);
-      this.filterShader.setUniform('canvasSize', [this.width, this.height]);
-      // filterParameter uniform only used for POSTERIZE, and THRESHOLD
-      // but shouldn't hurt to always set
-      this.filterShader.setUniform('filterParameter', filterParameter);
-      pg.rect(-this.width / 2, -this.height / 2, this.width, this.height);
+      fbo.draw(() => {
+        this._pInst.noStroke();
+        this._pInst.shader(this.filterShader);
+        this.filterShader.setUniform('tex0', target);
+        this.filterShader.setUniform('texelSize', texelSize);
+        this.filterShader.setUniform('canvasSize', [target.width, target.height]);
+        // filterParameter uniform only used for POSTERIZE, and THRESHOLD
+        // but shouldn't hurt to always set
+        this.filterShader.setUniform('filterParameter', filterParameter);
+        this._pInst.rect(-target.width / 2, -target.height / 2,
+          target.width, target.height);
+      });
 
     }
-    // draw pg contents onto main renderer
+    // draw fbo contents onto main renderer.
     this._pInst.push();
-    pg._pInst.resetMatrix();
     this._pInst.noStroke();
-    pg._pInst.imageMode(constants.CORNER);
-    pg._pInst.blendMode(constants.BLEND);
     this.clear();
     this._pInst.push();
     this.filterCamera._resize();
     this._pInst.setCamera(this.filterCamera);
     this._pInst.resetMatrix();
-    this._pInst.image(pg, -this.width / 2, -this.height / 2,
+    this._pInst.image(fbo, -this.width / 2, -this.height / 2,
```

Comment:
I think we still need these two lines above this call:
```js
this._pInst.imageMode(constants.CORNER);
this._pInst.blendMode(constants.BLEND);
```

Currently, if you change the image mode or blend mode, the filter gets drawn with an offset: https://editor.p5js.org/davepagurek/sketches/q442qjF_1

Reply:
Okay